Added wrappers for several key GTT functions

Yook74 opened the pull request:
- creates pixel
- clear all traces
- create rectangle
- create label
- update label
- set label font color
- set label font size
- set label background color
- load bitmap
- display bitmap
- set bitmap transparency
- initialize trace
- update trace
- create touch region
- create toggle region
- set GPO state
- setup animation
- load animation
- activate animation
- set animation frame
- get animation frame
- stop all animation
- resume all animation

Commit:
Added the following methods
- creates pixel
- clear all traces
- create rectangle
- create label
- update label
- set label font color
- set label font size
- set label background color
- load bitmap
- display bitmap
- set bitmap transparency
- initialize trace
- update trace
- create touch region
- create toggle region
- set GPO state
- setup animation
- load animation
- activate animation
- set animation frame
- get animation frame
- stop all animation
- resume all animation

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,411 @@ def update_bar_value(self, bar_id: IdType, value: int):
         self._receive_status_response(252, 105)
+
+    def create_pixel(self, x_pos: int, y_pos: int):
+        """Creates a bar graph which is really just a single bar.
```
Fix description

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,411 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        )
+
+    def draw_rectangle(self, x_pos: int, y_pos: int, width: int, height: int):
+        """Creates a rectangle which is really just a single bar.
```
Your docstring should include whether this is a filled-in rectangle or not

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,411 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param x_pos: the distance from the left edge of the screen in pixels
+        :param y_pos: the distance from the top edge of the screen in pixels
+        :param width: the width of the rectangle in pixels
+        :param height: the height of the bar in pixels
```
rectangle not bar

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,411 @@ def update_bar_value(self, bar_id: IdType, value: int):
+            ints_to_signed_shorts(x_pos, y_pos, width, height)
+        )
+
+    def create_a_label(self, label_id: int, x_pos: int, y_pos: int, width: int, height: int, font: int, R: int, G: int,
```
- probably change name to `create_label`
- Combine R, G, and B into a single hex string as I did in create_bar
- change `size` to `font_size`
- This is up for debate, but I don't think this needs to have a `value` parameter. There's a default value on the bar because the bar won't show up if it doesn't have a value, but in the case of a label I wouldn't expect it to show up without updating the value. Minimally, `value` should have a default value like `value='Label'`
- rot should have an enum and a docstring entry
- `label_id` should be an `IdType` not an `int`

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,411 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param R: the intensity of red, 0 to 255, used for label font color
+        :param G: the Intensity of green, 0 to 255, used for label font color
+        :param B: the Intensity of blue, 0 to 255, used for label font color
+        :param value: New ASCII formatted string to display within the label. String should be a single line in height.
```
This should support more than ASCII right? Also is that true that it won't tolerate a newline?

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,411 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        self.set_label_background_color(label_id, RB=0xFF, GB=0xFF, BB=0xFF)
+
+    def update_label(self, label_id: int, value: str):
+        """Updates string of the label identified by label_id
```
single line docstrings should take the format `"""docstring here"""`
Please check all your methods for correct type hints on the IDs and usage of `self._resove_id`

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,411 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        self.update_label(label_id, value)
+        self.set_label_font_color(label_id, R, G, B)
+        self.set_label_font_size(label_id, size)
+        self.set_label_background_color(label_id, RB=0xFF, GB=0xFF, BB=0xFF)
```
Why aren't these parameters? Look at how the bar handles foreground and background colors.

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,411 @@ def update_bar_value(self, bar_id: IdType, value: int):
+            value.encode('utf-8') + b'\0'
+        )
+
+    def set_label_font_color(self, label_id: int, R: int, G: int, B: int):
```
Needs to be a single hex string not three separate parameters
The docstring should mention what happens to text already in the label if you change he font color

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,411 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        self._receive_status_response(252, 21)
+
+    def set_label_font_size(self, label_id: int, size: int):
+        """Sets the font size of the label given by the size variable
```
"Sets the font size of the label given by **label_id**

Commit:
Date:3/17/22
Added the following methods
- creates pixel
- clear all traces
- create rectangle
- create label
- update label
- set label font color
- set label font size
- set label background color
- load bitmap
- display bitmap
- set bitmap transparency
- initialize trace
- update trace
- create touch region
- create toggle region
- set GPO state
- setup animation
- load animation
- activate animation
- set animation frame
- get animation frame
- stop all animation
- resume all animation

Date:3/18/22
Updated docstring in:
	create_pixel
	draw_rectangle

Updated create_a_label(create_label):
	Updated docstring
	Updated variables
		font size
		RGB(fg_color_hex) and RGB(bg_color_hex)

Updated set_label_font_color:
	Variable and docstring

Updated set_label_background_color:
	Variable and docstring

Updated id variable to IdType and _resolve_id method
	create label
	update label
	set label font color3
	set label font size
	set label background color
	load bitmap
	display bitmap
	set bitmap transparency
	initialize trace
	update trace
	create touch region
	create toggle region
	setup animation
	load animation

Added new method
	load_font necessary to use for other fonts and sizing

Commit:
fixed bug in exceptions

Commit:
Changed font_id to IdType

Commit:
Updated Trace style to use three different enums.

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
         self._receive_status_response(252, 105)
+
+    def create_pixel(self, x_pos: int, y_pos: int):
+        """Creates a pixel
```
This docstring is OK, but I want to comment that it it pretty easy to just reiterate yourself in the docstring instead of adding further explanation. That can easily lead to SNWB syndrome where the docstrings are largely useless:
![image](https://user-images.githubusercontent.com/25036872/159334245-4379066c-8eb3-4c19-bfef-eb6af5aef223.png)

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        )
+
+    def create_label(self, label_id: IdType, x_pos: int, y_pos: int, width: int, height: int, font: int,
+                     font_size: int, rot: int, fg_color_hex='FFFFFF', bg_color_hex='000000', value: str = "Label",
```
Lets add a default value of 0 for rot

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+    def create_label(self, label_id: IdType, x_pos: int, y_pos: int, width: int, height: int, font: int,
+                     font_size: int, rot: int, fg_color_hex='FFFFFF', bg_color_hex='000000', value: str = "Label",
+                     VJst: FontAlignVertical = FontAlignVertical.TOP,
+                     HJst: FontAlignHorizontal = FontAlignHorizontal.LEFT):
```
parameter names should be snake_cased. Please change these to `vertical_just` and `horizontal_just`

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param y_pos: the distance from the top edge of the screen in pixels
+        :param width: the width of the bar in pixels
+        :param height: the height of the bar in pixels
+        :param rot: the rotation of the text within the label.
```
Specify that it's in degrees

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+                     VJst: FontAlignVertical = FontAlignVertical.TOP,
+                     HJst: FontAlignHorizontal = FontAlignHorizontal.LEFT):
+        """Creates a label in a portion of the screen
+        :param label_id: index used to identify this label in the label list
```
Since this accepts either ints or strings, you should probably change this to more similar to the bar_id parameter descriptions I made

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param x_pos: the distance from the left edge of the screen in pixels
+        :param y_pos: the distance from the top edge of the screen in pixels
+        :param width: the width of the bar in pixels
+        :param height: the height of the bar in pixels
```
label not bar

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param rot: the rotation of the text within the label.
+        :param VJst: the vertical justification of text within the label
+        :param HJst: the horizontal justification of text within the label
+        :param font: the Font index of a previously loaded font to be used for the label.
```
Is at least one font loaded by default?

logver replied:
Yes it is in the font memory index is 0. if you want to add a new one you must load a new font and index the value. should I add this in the create label when its stated in the load_font method

Yook74 replied:
no, just give a default value of 0 for font and note that the default font doesn't support resizing

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param VJst: the vertical justification of text within the label
+        :param HJst: the horizontal justification of text within the label
+        :param font: the Font index of a previously loaded font to be used for the label.
+        :param fg_color_hex: a hex color string for the filled part of the label
```
This is for the text color, not for the "filled part of the label." Copying and pasting is sometimes appropriate but it can lead to lazy and misleading docstrings if you're not careful.

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param fg_color_hex: a hex color string for the filled part of the label
+        :param bg_color_hex: a hex color string for the unfilled part of the label
+        :param value: New UTF-8 string to display within the label. String should be a single line in height
+        :param font_size: Size of the font
```
In pixels?

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        )
+
+        self.update_label(label_id, value)
+        self.set_label_font_color(label_id, fg_color_hex)
```
why is the foreground color used twice?

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        self.set_label_font_size(label_id, font_size)
+        self.set_label_background_color(label_id, bg_color_hex)
+
+    def update_label(self, label_id: IdType, value: str = "Label"):
```
Remove the default value for `value`. It wouldn't make sense to call this function without filling that parameter.

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        self._conn.write(
+            bytes.fromhex('FE 11') +
+            label_id.to_bytes(1, 'big') +
+            b'\x02' +
```
for consistency, you might want to use `bytes.fromhex` here

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+            value.encode('utf-8') + b'\0'
+        )
+
+    def set_label_font_color(self, label_id: IdType, fg_color_hex='FFFFFF'):
```
Also probably remove the default value for `fg_color_hex` here

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        )
+
+    def set_label_font_color(self, label_id: IdType, fg_color_hex='FFFFFF'):
+        """Sets the font  color of the label given by the RGB
```
this seems unfinished

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+    def set_label_font_color(self, label_id: IdType, fg_color_hex='FFFFFF'):
+        """Sets the font  color of the label given by the RGB
+        :param label_id: index used to identify this label in the label list
```
see above comment about rephrasing ID docstrings to include strings or ints

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+    def set_label_font_color(self, label_id: IdType, fg_color_hex='FFFFFF'):
+        """Sets the font  color of the label given by the RGB
+        :param label_id: index used to identify this label in the label list
+        :param fg_color_hex: a hex color string for the filled part of the label
```
see above comment about copy pasting

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+    def set_label_font_size(self, label_id: IdType, size: int):
+        """Sets the font size of the label given by the label_id variable
+        :param label_id: index used to identify this label in the label list
+        :param size: Size of the font
```
- Mention that the default font doesn't support resizing
- see above comments about font size units and id descriptions

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        """""Set the background color of an existing label by the RB, GB, and BB
+        :param label_id: index used to identify this label in the label list
+        :param bg_color_hex: a hex color string for the unfilled part of the label
+        """
```
- remove extra `""`
- remove default value for parameter
- fix docstring

logver replied:
fix this with set_label_font_color too

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        """Load a font file from the SD card into a font buffer for use.
+        :param font_id: Index used to identify the font. Has to be an int.
+        :param fileName: filename, and path from the root folder, of the animation file to load.
+        """
```
- snake_case `fileName`
- fix ID docstring
- mention that this is a tff file
- give an example fileName

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+    def load_bitmap(self, bitmap_id: IdType, fileName: str):
+        """Load a bitmap file from the SD card into a bitmap buffer for use.
+        :param bitmap_id: index used to identify the bitmap. Specific to bitmaps, and screen rectangles
+        :param fileName: filename, and path from the root folder, of the bitmap file to load
```
- snake_case `fileName`
- give example `fileName`
- fix ID docstring

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+        self._receive_status_response(252, 95)
+
+    def display_bitmap(self, bitmap_id: IdType, x_pos: int, y_pos: int):
```
Lets add a convenience function which loads a bitmap and displays it since that is usually going to be what people want to do. That function should just call these two functions.

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+    def display_bitmap(self, bitmap_id: IdType, x_pos: int, y_pos: int):
+        """Display a bitmap image on the screen, from the bitmap buffer
+        :param bitmap_id: Index used to identify the desired file in the bitmap buffer
```
fix id docstring

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+    def set_bitmap_transparency(self, bitmap_id: IdType, fg_color_hex='FFFFFF'):
+        """Set the transparent color for all future renderings of a specific bitmap index. Does not affect previously
+        drawn versions of the specified bitmap
```
This docstring is a good start, but can you expand on "Set the transparent color?" Maybe give an example. Also fg_color_hex is not the right name for this parameter, and it shouldn't have a default value.

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        self._receive_status_response(252, 98)
+
+    def initialize_trace(self, trace_id: IdType, x_pos: int, y_pos: int, width: int, height: int,
+                         min_value: int, max_value: int, step: int, value: int, fg_color_hex='FFFFFF',
```
- consider adding default values for some of these parameters
- do we not get to set the background color?

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+    def update_trace(self, trace_id: IdType, value: int):
+        """Update the value of the trace at the specified index.
+        :param trace_id: Index used to identify this trace in the trace list
```
rephrase to include string IDs

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+            trace_id.to_bytes(1, 'big') +
+            ints_to_signed_shorts(value)
+
+        )
```
- remove extra space
- No response from this one? all the other update commands have responses

logver replied:
No response

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+    def create_touch_region(self, region_id: IdType, x_pos: int, y_pos: int, width: int, height: int,
+                            up: int, down: int):
+        """Create a region of the screen that responds to touch events with a unique message and momentary visual update
```
does it respond with a unique message? I know the docs say it does but is that real?

logver replied:
No, it returns the same message b'\xfc\x87\x00\x02\x00\x03' no matter size or position

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        )
+
+    def create_touch_region(self, region_id: IdType, x_pos: int, y_pos: int, width: int, height: int,
+                            up: int, down: int):
```
change `up` and `down` to `up_bitmap_id` and `down_bitmap_id` and change them to IdType type hints

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param x_pos: the distance from the left edge of the screen in pixels
+        :param y_pos: the distance from the top edge of the screen in pixels
+        :param width: width of the trace region
+        :param height: height of the trace region
```
this is not a trace, watch the copy-paste

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+            ints_to_signed_shorts(x_pos, y_pos) +
+            ints_to_unsigned_shorts(width, height) +
+            up.to_bytes(1, 'big') +
+            down.to_bytes(1, 'big')
```
resolve up and down IDs

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        while True:
+            response = self._conn.read(6)  # binary value
+            if response == b'\xfc\x87\x00\x02\x00\x03':
+                print('Contact was made')
```
what is going on here

logver replied:
We'll talk about this

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+    def create_toggle_region(self, region_id: IdType, x_pos: int, y_pos: int, width: int, height: int,
+                             off_id: int, on_id: int):
+        """Create a region of the screen that responds to touch events with a unique message and toggleable visual
```
Once again, what does "responds with a unique message" actually mean?

logver replied:
create touch region: b'\xfc\x87\x00\x02\x00\x03' this is the unique message
create toggle region: (252, 150)

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param width: width of the trace region
+        :param height: height of the trace region
+        :param off_id: index of the loaded bitmap displayed when the region is in an inactive state
+        :param on_id: index of the loaded bitmap displayed when the region is in a toggled state
```
same comments as touch_region docstring

logver replied:
Did you want me to change the on and off id comment because it is different

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        """Toggle the specified General Purpose Output pin on or off, sourcing up to 15mA current at
+        5V per GPO or sinking to ground. This command can be used to control devices, or signal a host device.
+        :param pin: GPO to be controlled
+        :param state: GPO state, as per eGPOSetting Values
```
`"""Set the specified General Purpose Output pin on or off.
Sourcing up to 15mA of current at 5V per pin or sinking to ground.
"""`
No need for parameter docstrings, it's pretty clear what they are

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+        self._receive_status_response(252, 150)
+
+    def set_GPO_state(self, pin: int, state: bool):
```
lowercase `GPO` in the function name even though it's an acronym

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param x_pos: the distance from the left edge of the screen in pixels
+        :param y_pos: the distance from the top edge of the screen in pixels
+        :param fileName: filename, and path from the root folder, of the animation file to load.
+        :param play: desired animation state
```
- lets change the first two parameters to `memory_id: IdType,  display_id: IdType`
- why are we specifying a `fileName`?
- The docstring could be clearer

logver replied:
What is memory_id: IdType, display_id: IdType replacing?

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+            ints_to_signed_shorts(x_pos, y_pos)
+        )
+
+        self.load_animation(animation_id, fileName, animation_instance, play)
```
why is this here?

logver replied:
Lets talk about this.

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        finally:
+            self._conn.timeout = prev_timeout
+
+        self.activate_animation(animation_instance, play)
```
why is this here?

logver replied:
Lets talk about this.

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        :param fileName: filename, and path from the root folder, of the animation file to load.
+        :param animation_instance: index used to identify this animation instance in the animation list.
+        :param play: desired animation state
+        """
```
- This function should not take `animation_instance` and `play` as parameters
- snake_case `fileName`
- give a tutorial of how to create an animation file in the docstring
- note that this can take a little while to complete

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+    def activate_animation(self, animation_instance: int, play=True):
+        """
+        :param animation_instance: index used to identify this animation instance in the animation list
```
change to `display_id: IdType`

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        """Set the current frame of a displayed animation. If the frame exceeds the total number present, the animation
+        will be set to the first frame.
+        :param animation_instance: index used to identify this animation instance in the animation list
+        :param frame: Number of the frame to be displayed. Needs to be less that actual amount ex: input 7 for 8 photos
```
That means it's zero-based, right?

logver replied:
Yes it is

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        will be set to the first frame.
+        :param animation_instance: index used to identify this animation instance in the animation list
+        :param frame: Number of the frame to be displayed. Needs to be less that actual amount ex: input 7 for 8 photos
+        :animation must be loaded and stopped before setting frame
```
put this at the beginning of the docstring and omit the leading `:`

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+            bytes.fromhex('FE  C4') +
+            animation_instance.to_bytes(1, 'big')
+        )
+        response = self._conn.read(6)
```
use `receive_query_response`

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+    def get_animation_frame(self, animation_instance: int):
+        """Gets the current frame of an existing animation instance
+        :param animation_instance: index used to identify this animation instance in the animation list
```
change to `display_id: IdType`

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+    def stop_all_animation(self):
+        """Stops all currently running animation instances at their present frame
+        """
```
`"""single line docstrings look like this"""`

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+        )
+
+    def resume_all_animation(self):
+        """Resumes all stopped animation instances from their present frame.
```
`"""single line docstrings look like this"""`

Yook74 commented on gtt/gtt_display.py:
```diff
@@ -182,3 +182,417 @@ def update_bar_value(self, bar_id: IdType, value: int):
+
+        self.activate_animation(animation_instance, play)
+
+    def activate_animation(self, animation_instance: int, play=True):
```
Lets add a new convenience function that loads the animation, sets it up, and plays it called `load_and_play_animation`

Yook74 reviewed (commented):
Lots of little things to fix

logver committed:
Updated docstrings
Added IdType for all variable ID
Updated RBG to fg_color_hex and bg_color_hex

To do for Andrew
bitmap transparency
response to toggle and touch region

yook-at-work committed:
Moved the logic of picking an unused ID into a separate method and used it to fix load_and_play_animation

Yook74 committed:
Added method to enter mass storage mode

Yook74 committed:
Discovered the true behavior of bitmap transparency

Yook74 committed:
Docstring and parameter name updates

Yook74 committed:
Added method to set drawing color

Yook74 committed:
Added new methods to auto doc generation

Commit:
Added
test_animation.py
test_bitmap.py
test_draw.py
test_label.py

Have been tested with GTT35A

Add important files used to test
Files added:
Animation folder with pictures and txt file
megaman folder with pictures and txt file
lorge folder with font file
megaman.bmp
Logo35A.bmp

Commit:
Updated
fixtures.py
gtt_display.py
test_bitmap.py
test_animation.py

yook-at-work committed:
removed "Files for testing" and replaced with tests/assets which has only things compatible with our license

yook-at-work committed:
Added animation asset files

Yook74 committed:
Added Logan to the authors list

Yook74 committed:
Merge remote-tracking branch 'origin/MVP' into MVP

Yook74 committed:
Figured out some slightly intuitive parameters in the trace creation method

I also removed the call to update_trace in that method because the trace always starts at 0 whether we like it or not

Yook74 committed:
attempting to re-use an ID now results in an IdConflictError instead of a ValueError

Yook74 committed:
fixed label fonts

Yook74 committed:
draw_rectangle can now be filled or unfilled
